### DISCUSSION TITLE
Lifecycle hooks

### DISCUSSION THREAD

**Discussion by ceilfors** [nearst/laconia#927] [Ideas]
Recently I've got a use case of trying to execute something before a Lambda times out. This made me think of a lifecycle that I should have been able to listen to, for example, it may look like this:

```js
laconia(app)
  .register(...)
  .on('30sBeforeTimeout', () => { /* do something * / })
```

**Discussion**

- Is this something that we should introduce?
- What other lifecycle hooks that developers may find useful here?
- Should the API be using .on(string)? I felt that a string makes it too easy for developers to have a typo. How might the API look like?

**Comment by hugosenari** [comment]
Maybe something like:
* 'error': to works like try/catch for unhandled errors;
* 'finally': to works like try/catch/finally;
* 'beforeHandle': to set adapters to event;

**Comment by ceilfors** [comment]
@hugosenari Thanks for joining this discussion! On try/catch or setting control flow, at the moment one of the philosophy I'm trying to encourage is [use JavaScript as close as possible](https://laconiajs.io/docs/introduction/philosophy#design-for-developer-experience). Control flow like try/catch is a JavaScript syntax, I think a middleware can be introduced to tackle this. I think it's quite similar to `beforeHandle` for adapter, where high order function is used instead. What do you think?

**Comment by hugosenari** [comment]
Maybe we could implement everything as [HOF adapter](https://gist.github.com/hugosenari/ee18f739bce8f96c41b5904d731184fb) and see what makes sense move to core.

```js
const myAdapter = hofAdapter({
  mappings: [getBody],
  onEnd: result => result,
  onSuccess: identity,
  throwMillisBeforeTimeout: 500,
  onError: error => { throw error; }
});

exports.handler = laconia(myAdapter(body => ({ data: body })));
```

Until we could do this:

```js
module.exports = laconia(body => ({ data: body }))
  .on('end', () => {})
  .on('mapping', getBody)
  .on('success', result => result)
  .on('error', error => { throw error; )
  .on('beforeTimeout': timeroutErrorHandler);
```
